File: Perl/src/collatz.pl
#!/usr/bin/perl

use strict;
use warnings FATAL => 'all';

sub collatz_iteration {
    my ($number) = @_;
    my $steps = 0;

    while ($number != 1) {
        print($number . " ");
        if ($number % 2 == 0) {
            $number = $number / 2;
        }
        else {
            $number = 3 * $number + 1;
        }
        $steps++;
    }
    print($number . "\n");
    return ($steps);
}

sub collatz_recursive {
    no warnings 'recursion';
    my ($number) = @_;

    print($number . " ");
    if ($number == 1) {
        return ();
    }
    if ($number % 2 == 0) {
        collatz_recursive($number / 2);
    }
    else {
        collatz_recursive(3 * $number + 1);
    }
}

my $input_number = 27;
print(collatz_iteration($input_number) . "\n");
collatz_recursive($input_number);

File: Go/src/collatz.go
package main

import "fmt"

func collatzIteration(number int) int {
	steps := 0
	for number != 1 {
		fmt.Printf("%d ", number)
		if (number % 2) == 0 {
			number = number / 2
		} else {
			number = 3*number + 1
		}
		steps = steps + 1
	}
	fmt.Printf("%d\n", number)
	return steps
}

func collatzRecursive(number int) {
	fmt.Printf("%d ", number)
	if number == 1 {
		return
	}
	if number%2 == 0 {
		collatzRecursive(number / 2)
	} else {
		collatzRecursive(3*number + 1)
	}
}

func main() {
	inputNumber := 27
	fmt.Printf("%d\n", collatzIteration(inputNumber))
	collatzRecursive(inputNumber)
}
